server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContent.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+/**
+ * An extension of {@link ToXContent} that can be serialized in chunks by creating a {@link ChunkedXContentSerialization}.
+ * This is used by the REST layer to implement flow control that does not rely on blocking the serializing thread when writing the
+ * serialized bytes to a non-blocking channel.
+ */
+public interface ChunkedToXContent extends ToXContent {
+
+    /**
+     * Create a serialization session for the implementation instance that will flush to the given {@code builder} and use the given
+     * {@code params} the same way {@link ToXContent#toXContent(XContentBuilder, Params)} would.
+     * @param builder builder to flush to
+     * @param params params for serialization
+     * @return serialization session
+     */
+    ChunkedXContentSerialization toXContentChunked(XContentBuilder builder, ToXContent.Params params);
```

Comment:
I wonder if we should instead add a method that returns the chunks as iterator or stream over`ToXContent` (or a new interface if there are compelling reasons). The individual chunks could then also be chunked again.

I think this will simplify the logic in the code that implements this, since they no longer need booleans for whether a chunk has been written or not (`wroteStart`) and also do not need to handle the nested chunking.

Reply:
That's what I had at first. What I really disliked about that was, that you'd have to pass the params in over and over when they never change across chunks, so I felt it was cleaner to make this clear via the interface.
I don't really think we'd save any complexity with this move anyway, you'd still have to build custom iterators for everything? You'll still have to have nested chunking in so far that one iterator will have to have logic for when to delegate to another one or when to return its own the last chunk etc?

Last but not least, using `Iterator` and returning some `ToXContent` might behave quite a bit worse performance wise because it won't compile as nicely as having all of this internal to the implementation (won't matter if chunks are large, but will matter more if they are small and it's not always trivial to predict chunk size).
-> I kinda like this better tbh.